feat: initialize projenrc.js with commented options

Chriscbr opened the pull request:
*Issue #, if available:* Closes #168 

*Description of changes:* Updated default rendering function for .projenrc.js config file so that it gets initialize with a list of all options available based on the project type.

Also updated the docstrings of several props so that the "default" field would be a proper value.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Chriscbr commented:
This feature was easier than expected to implement. The only issue right now not all props seem to show up, and I'm honestly not sure why... For example if I run `npx projen new typescript`, around 50 or so of the possible config options all appear in alphabetical order (like shown below), but all of the TypeScript-specific options don't seem to appear. I might have to dig a little deeper into how JSII picks up these labels (because these properties are still appearing correctly in API.md).

```ts
const project = new NodeProject({
  // allowLibraryDependencies: true,                                           /** Allow the project to include `peerDependencies` and `bundledDependencies`. */
  // antitamper: true,                                                         /** Checks that after build there are no modified files on git. */
  // authorEmail: undefined,                                                   /** Author's e-mail. */
  ...
```

eladb commented on src/node-project.ts:
```diff
@@ -69,7 +69,7 @@ export interface NodeProjectCommonOptions {
    * Version of projen to install.
    *
-   * @default - latest version
+   * @default - Semver.latest()
```
```suggestion
   * @default Semver.latest()
```

Chriscbr commented on src/jsii-project.ts:
```diff
@@ -87,7 +87,7 @@ export interface JsiiProjectOptions extends NodeProjectCommonOptions {
   /**
    * Jest options
-   * @default - defaults
+   * @default {}
```
I think I can do this fairly reasonably - using JSON.parse I can check if the string is any kind of normal javascript value (which covers most defaults). The only case where this would be an issue are default options where the default is a value from a class in the library, e.g. `Semver.latest()` or `MyEnum.OptionA`. I think the easiest way to work around this limitation is just to pretend its a string (so the docstring would say instead say `@default "Semver.latest()"`).

eladb replied:
Yeah, I think that's a reasonable approximation

Chriscbr commented on src/cli/cmds/new.ts:
```diff
@@ -104,9 +104,36 @@ function generateProjenConfig(baseDir: string, type: inventory.ProjectType, para
-function renderParams(params: any) {
-  return JSON.stringify(params, undefined, 2)
-    .replace(/\"(.*)\":/g, '$1:'); // remove quotes from field names
+function makePadding(paddingLength: number): string {
```
In general I'm adding padding in the middle of strings in the code below, so I think the helper function is still necessary. That said you raise a good point about efficiency, and I think using Array.join isn't really the fastest way to create a string - I just did a quick benchmark and I think String.repeat is faster.

eladb replied:
Yes, `' '.repeat(paddingLength+1)` should do the trick.

eladb commented:
looking forward for another iteration

eladb commented on src/jsii-project.ts:
```diff
@@ -133,7 +133,8 @@ export interface JsiiPythonTarget {
    * The registry url to use when releasing packages.
-   * @default - twine default
+   *
+   * By default uses the default twine registry url.
```
I think we still need @default here but with a "behavioral" description. The syntax we use in the CDK for this is to use `-` as the value

```
@default - DESCRIPTION
```

Chriscbr replied:
Ahhh okay - so the kind of metadata JSII spits out to us looks like this:
```json
{
    "abstract":true,
    "docs":{
        "default":"- true",
        "remarks":"By default is true if mergify auto-merge is enabled (default).",
        "stability":"experimental",
        "summary":"Automatically merge projen upgrade PRs when build passes. Applies the `mergifyAutoMergeLabel` to the PR if enabled."
    },
    "immutable":true,
    "locationInModule":{
        "filename":"src/node-project.ts",
        "line":390
    },
    "name":"projenUpgradeAutoMerge",
    "optional":true,
    "type":{
        "primitive":"boolean" // or e.g. "fqn": "projen.Semver"
    }
},
```
and it gets processed as
```json
{
   "path":[
      "projenUpgradeAutoMerge"
   ],
   "name":"projenUpgradeAutoMerge",
   "docs":"Automatically merge projen upgrade PRs when build passes. Applies the `mergifyAutoMergeLabel` to the PR if enabled.",
   "type":"boolean",
   "switch":"projen-upgrade-auto-merge",
   "default":"true",
   "optional":true
}
```
I didn't realize projen was stripping out the "-". Based on that, I've added a flag that indicates whether the default value is a description (based on this convention), which allows us to allow any kinds of defaults like these.

eladb commented on src/node-project.ts:
```diff
@@ -149,7 +149,7 @@ export interface NodeProjectCommonOptions {
    * Version of projen to install.
    *
-   * @default - latest version
+   * @default "Semver.latest()"
```
Can't say I am really fond of this quirk. Can we use a bit of heuristic to allow users to specify such default values without strings. 

Since we know the type of the property, we can be smarter about validation of the default.

For example, we can check if the value starts with the type name (if it's not a string)

eladb reviewed (commented):
A few additional comments. 

Also, can you post a comment with an example snippet of how e.g. a new jsii project looks like after this change?

Chriscbr commented:
Here's what .projenrc.js looks like after creating a new jsii project with the features on this PR:

```
const { JsiiProject } = require('projen');

const project = new JsiiProject({
  // allowLibraryDependencies: true,                                           /** Allow the project to include `peerDependencies` and `bundledDependencies`. */
  // antitamper: true,                                                         /** Checks that after build there are no modified files on git. */
  authorAddress: "example@example.com",                                        /** Email or URL of the library author. */
  // authorEmail: undefined,                                                   /** undefined */
  authorName: "Chriscbr",                                                      /** The name of the library author. */
  // authorOrganization: undefined,                                            /** undefined */
  // authorUrl: undefined,                                                     /** undefined */
  // autoDetectBin: true,                                                      /** Automatically add all executables under the `bin` directory to your `package.json` file under the `bin` section. */
  // bin: undefined,                                                           /** Binary programs vended with your module. */
  // buildWorkflow: true,                                                      /** Define a GitHub workflow for building PRs. */
  // bundledDependencies: undefined,                                           /** undefined */
  // bundledDeps: undefined,                                                   /** List of dependencies to bundle into this module. */
  // compat: false,                                                            /** Automatically run API compatibility test against the latest version published to npm after compilation. */
  // compatIgnore: ".compatignore",                                            /** Name of the ignore file for API compatibility tests. */
  // copyrightOwner: undefined,                                                /** License copyright owner. */
  // copyrightPeriod: undefined,                                               /** The copyright years to put in the LICENSE file. */
  // defaultReleaseBranch: "master",                                           /** The name of the main release branch. */
  // dependabot: true,                                                         /** Include dependabot configuration. */
  // dependabotOptions: undefined,                                             /** Options for dependabot. */
  // dependencies: undefined,                                                  /** undefined */
  // deps: [],                                                                 /** Runtime dependencies of this module. */
  // description: undefined,                                                   /** Library description. */
  // devDependencies: undefined,                                               /** undefined */
  // devDeps: [],                                                              /** Build dependencies for this module. */
  // docgen: true,                                                             /** Automatically generate API.md from jsii. */
  // dotnet: undefined,                                                        /** undefined */
  // entrypoint: "lib/index.js",                                               /** Module entrypoint (`main` in `package.json`). */
  // eslint: true,                                                             /** Install eslint. */
  // java: undefined,                                                          /** undefined */
  // jest: true,                                                               /** Use jest for unit tests. */
  // jestOptions: undefined,                                                   /** Jest options. */
  // jsiiVersion: undefined,                                                   /** undefined */
  // keywords: undefined,                                                      /** Keywords to include in `package.json`. */
  // libdir: "lib",                                                            /** Compiler artifacts output directory. */
  // license: undefined,                                                       /** undefined */
  // maxNodeVersion: undefined,                                                /** Minimum node.js version to require via `engines` (inclusive). */
  // mergify: true,                                                            /** Adds mergify configuration. */
  // mergifyAutoMergeLabel: "auto-merge",                                      /** Automatically merge PRs that build successfully and have this label. */
  // mergifyOptions: undefined,                                                /** Options for mergify. */
  // minNodeVersion: undefined,                                                /** Minimum Node.js version to require via package.json `engines` (inclusive). */
  name: "projen-test",                                                         /** The name of the library. */
  // npmDistTag: "latest",                                                     /** The dist-tag to use when releasing to npm. */
  // npmignore: undefined,                                                     /** Additional entries to .npmignore. */
  // npmignoreEnabled: true,                                                   /** Defines an .npmignore file. Normally this is only needed for libraries that are packaged as tarballs. */
  // npmRegistry: "registry.npmjs.org",                                        /** The registry url to use when releasing packages. */
  // packageManager: NodePackageManager.YARN,                                  /** The Node Package Manager used to execute scripts. */
  // peerDependencies: undefined,                                              /** undefined */
  // peerDependencyOptions: undefined,                                         /** Options for `peerDeps`. */
  // peerDeps: [],                                                             /** Peer dependencies for this module. */
  // projenDevDependency: true,                                                /** Indicates of "projen" should be installed as a devDependency. */
  // projenUpgradeAutoMerge: undefined,                                        /** Automatically merge projen upgrade PRs when build passes. */
  // projenUpgradeSchedule: [ "0 6 * * *" ],                                   /** Customize the projenUpgrade schedule in cron expression. */
  // projenUpgradeSecret: undefined,                                           /** Periodically submits a pull request for projen upgrades (executes `yarn projen:upgrade`). */
  // projenVersion: Semver.latest(),                                           /** Version of projen to install. */
  // pullRequestTemplate: true,                                                /** Include a GitHub pull request template. */
  // pullRequestTemplateContents: undefined,                                   /** The contents of the pull request template. */
  // python: undefined,                                                        /** undefined */
  // releaseBranches: [ "master" ],                                            /** Branches which trigger a release. */
  // releaseEveryCommit: true,                                                 /** Automatically release new versions every commit to one of branches in `releaseBranches`. */
  // releaseSchedule: undefined,                                               /** CRON schedule to trigger new releases. */
  // releaseToNpm: false,                                                      /** Automatically release to npm when new versions are introduced. */
  // releaseWorkflow: true,                                                    /** Define a GitHub workflow for releasing from "master" when new versions are bumped. */
  repository: "https://github.com/Chriscbr/projen-test.git",                   /** Git repository URL. */
  // rootdir: ".",                                                             /** undefined */
  // scripts: {},                                                              /** npm scripts to include. */
  // srcdir: "src",                                                            /** Typescript sources directory. */
  // stability: undefined,                                                     /** undefined */
  // start: true,                                                              /** Defines a `yarn start` interactive experience. */
  // startOptions: undefined,                                                  /** Options for `yarn start`. */
  // testdir: "test",                                                          /** Tests directory. */
  // workflowBootstrapSteps: "yarn install --frozen-lockfile && yarn projen",  /** Workflow steps to use in order to bootstrap this repo. */
  // workflowContainerImage: undefined,                                        /** Container image to use for GitHub workflows. */
  // workflowNodeVersion: undefined,                                           /** The node version to use in GitHub workflows. */
});

project.synth();
```

Note how there is `// projenVersion: Semver.latest()` which corresponds to `@default Semver.latest()` from its docstring, and `// copyrightPeriod: undefined` which corresponds to `@default - current year` from its docstring.

hoegertn commented:
Can we put this behind a cli flag or at least add a flag to disable this?
I think this is pretty unreadable for some project types with tens of options. 
We have auto-completion to know what possible options we can set imho.

Chriscbr commented:
Great idea - I've gone ahead and added a CLI flag, `--removeComments`. I've opted to have the comments show by default since it's most helpful for new users, and it only gets generated once.

I've also added a minor quality of life improvement by organizing the options according to the interface they are from, to break down the big list a little bit. (For example, in a ReactProject, there a section appears for NodeProjectCommonOptions, a section for NodeProjectOptions, a section for TypeScript options, and if there were custom react options, then ReactTypeScriptProjectOptions). Here is how it looks with jsii:

```
const { JsiiProject } = require('projen');

const project = new JsiiProject({
  /* NodeProjectCommonOptions */
  // allowLibraryDependencies: true,                                           /* Allow the project to include `peerDependencies` and `bundledDependencies`. */
  // antitamper: true,                                                         /* Checks that after build there are no modified files on git. */
  // autoDetectBin: true,                                                      /* Automatically add all executables under the `bin` directory to your `package.json` file under the `bin` section. */
  // bin: undefined,                                                           /* Binary programs vended with your module. */
  // buildWorkflow: true,                                                      /* Define a GitHub workflow for building PRs. */
  // bundledDependencies: undefined,                                           /* undefined */
  // bundledDeps: undefined,                                                   /* List of dependencies to bundle into this module. */
  // copyrightOwner: undefined,                                                /* License copyright owner. */
  // copyrightPeriod: undefined,                                               /* The copyright years to put in the LICENSE file. */
  // defaultReleaseBranch: "master",                                           /* The name of the main release branch. */
  // dependabot: true,                                                         /* Include dependabot configuration. */
  // dependabotOptions: undefined,                                             /* Options for dependabot. */
  // dependencies: undefined,                                                  /* undefined */
  // deps: [],                                                                 /* Runtime dependencies of this module. */
  // devDependencies: undefined,                                               /* undefined */
  // devDeps: [],                                                              /* Build dependencies for this module. */
  // entrypoint: "lib/index.js",                                               /* Module entrypoint (`main` in `package.json`). */
  // keywords: undefined,                                                      /* Keywords to include in `package.json`. */
  // libdir: "lib",                                                            /* Compiler artifacts output directory. */
  // maxNodeVersion: undefined,                                                /* Minimum node.js version to require via `engines` (inclusive). */
  // mergify: true,                                                            /* Adds mergify configuration. */
  // mergifyAutoMergeLabel: "auto-merge",                                      /* Automatically merge PRs that build successfully and have this label. */
  // mergifyOptions: undefined,                                                /* Options for mergify. */
  // minNodeVersion: undefined,                                                /* Minimum Node.js version to require via package.json `engines` (inclusive). */
  // npmDistTag: "latest",                                                     /* The dist-tag to use when releasing to npm. */
  // npmignore: undefined,                                                     /* Additional entries to .npmignore. */
  // npmignoreEnabled: true,                                                   /* Defines an .npmignore file. Normally this is only needed for libraries that are packaged as tarballs. */
  // npmRegistry: "registry.npmjs.org",                                        /* The registry url to use when releasing packages. */
  // packageManager: NodePackageManager.YARN,                                  /* The Node Package Manager used to execute scripts. */
  // peerDependencies: undefined,                                              /* undefined */
  // peerDependencyOptions: undefined,                                         /* Options for `peerDeps`. */
  // peerDeps: [],                                                             /* Peer dependencies for this module. */
  // projenDevDependency: true,                                                /* Indicates of "projen" should be installed as a devDependency. */
  // projenUpgradeAutoMerge: undefined,                                        /* Automatically merge projen upgrade PRs when build passes. */
  // projenUpgradeSchedule: [ "0 6 * * *" ],                                   /* Customize the projenUpgrade schedule in cron expression. */
  // projenUpgradeSecret: undefined,                                           /* Periodically submits a pull request for projen upgrades (executes `yarn projen:upgrade`). */
  // projenVersion: Semver.latest(),                                           /* Version of projen to install. */
  // pullRequestTemplate: true,                                                /* Include a GitHub pull request template. */
  // pullRequestTemplateContents: undefined,                                   /* The contents of the pull request template. */
  // releaseBranches: [ "master" ],                                            /* Branches which trigger a release. */
  // releaseEveryCommit: true,                                                 /* Automatically release new versions every commit to one of branches in `releaseBranches`. */
  // releaseSchedule: undefined,                                               /* CRON schedule to trigger new releases. */
  // releaseToNpm: false,                                                      /* Automatically release to npm when new versions are introduced. */
  // releaseWorkflow: true,                                                    /* Define a GitHub workflow for releasing from "master" when new versions are bumped. */
  // scripts: {},                                                              /* npm scripts to include. */
  // srcdir: "src",                                                            /* Typescript sources directory. */
  // start: true,                                                              /* Defines a `yarn start` interactive experience. */
  // startOptions: undefined,                                                  /* Options for `yarn start`. */
  // testdir: "test",                                                          /* Tests directory. */
  // workflowBootstrapSteps: "yarn install --frozen-lockfile && yarn projen",  /* Workflow steps to use in order to bootstrap this repo. */
  // workflowContainerImage: undefined,                                        /* Container image to use for GitHub workflows. */
  // workflowNodeVersion: undefined,                                           /* The node version to use in GitHub workflows. */

  /* JsiiProjectOptions */
  authorAddress: "yourname@example.com",                                       /* Email or URL of the library author. */
  // authorEmail: undefined,                                                   /* undefined */
  authorName: "Chriscbr",                                                      /* The name of the library author. */
  // authorOrganization: undefined,                                            /* undefined */
  // authorUrl: undefined,                                                     /* undefined */
  // compat: false,                                                            /* Automatically run API compatibility test against the latest version published to npm after compilation. */
  // compatIgnore: ".compatignore",                                            /* Name of the ignore file for API compatibility tests. */
  // description: undefined,                                                   /* Library description. */
  // docgen: true,                                                             /* Automatically generate API.md from jsii. */
  // dotnet: undefined,                                                        /* undefined */
  // eslint: true,                                                             /* Install eslint. */
  // java: undefined,                                                          /* undefined */
  // jest: true,                                                               /* Use jest for unit tests. */
  // jestOptions: undefined,                                                   /* Jest options. */
  // jsiiVersion: undefined,                                                   /* undefined */
  // license: undefined,                                                       /* undefined */
  name: "projen-test",                                                         /* The name of the library. */
  // python: undefined,                                                        /* undefined */
  repository: "https://github.com/Chriscbr/projen-test.git",                   /* Git repository URL. */
  // rootdir: ".",                                                             /* undefined */
  // stability: undefined,                                                     /* undefined */
});

project.synth();
```

eladb reviewed (changes requested):
Looks good! A few more ergonomics suggestions:

1. Hoist all the assigned options to the top (so it's easy to just remove the comments)
2. Remove the deprecated options
3. Alpha sort

pgollucci commented:
> Looks good! A few more ergonomics suggestions:
> 
> 1. Hoist all the assigned options to the top (so it's easy to just remove the comments)
> 2. Remove the deprecated options
> 3. Alpha sort

Agree. This is cool!

Chriscbr commented:
Updated with all three suggestions:

```ts
const { JsiiProject } = require('projen');

const project = new JsiiProject({
  authorAddress: "yourname@email.com",
  authorName: "Chriscbr",
  name: "projen-test",
  repository: "https://github.com/Chriscbr/projen-test.git",

  /* JsiiProjectOptions */
  // authorOrganization: undefined,                                            /* undefined */
  // compat: false,                                                            /* Automatically run API compatibility test against the latest version published to npm after compilation. */
  // compatIgnore: ".compatignore",                                            /* Name of the ignore file for API compatibility tests. */
  // description: undefined,                                                   /* Library description. */
  // docgen: true,                                                             /* Automatically generate API.md from jsii. */
  // dotnet: undefined,                                                        /* undefined */
  // eslint: true,                                                             /* Install eslint. */
  // java: undefined,                                                          /* undefined */
  // jest: true,                                                               /* Use jest for unit tests. */
  // jestOptions: undefined,                                                   /* Jest options. */
  // jsiiVersion: undefined,                                                   /* undefined */
  // license: undefined,                                                       /* undefined */
  // python: undefined,                                                        /* undefined */
  // rootdir: ".",                                                             /* undefined */
  // stability: undefined,                                                     /* undefined */

  /* NodeProjectCommonOptions */
  // allowLibraryDependencies: true,                                           /* Allow the project to include `peerDependencies` and `bundledDependencies`. */
  // antitamper: true,                                                         /* Checks that after build there are no modified files on git. */
  // autoDetectBin: true,                                                      /* Automatically add all executables under the `bin` directory to your `package.json` file under the `bin` section. */
  // bin: undefined,                                                           /* Binary programs vended with your module. */
  // buildWorkflow: true,                                                      /* Define a GitHub workflow for building PRs. */
  // bundledDeps: undefined,                                                   /* List of dependencies to bundle into this module. */
  // copyrightOwner: undefined,                                                /* License copyright owner. */
  // copyrightPeriod: undefined,                                               /* The copyright years to put in the LICENSE file. */
  // defaultReleaseBranch: "master",                                           /* The name of the main release branch. */
  // dependabot: true,                                                         /* Include dependabot configuration. */
  // dependabotOptions: undefined,                                             /* Options for dependabot. */
  // deps: [],                                                                 /* Runtime dependencies of this module. */
  // devDeps: [],                                                              /* Build dependencies for this module. */
  // entrypoint: "lib/index.js",                                               /* Module entrypoint (`main` in `package.json`). */
  // keywords: undefined,                                                      /* Keywords to include in `package.json`. */
  // libdir: "lib",                                                            /* Compiler artifacts output directory. */
  // maxNodeVersion: undefined,                                                /* Minimum node.js version to require via `engines` (inclusive). */
  // mergify: true,                                                            /* Adds mergify configuration. */
  // mergifyAutoMergeLabel: "auto-merge",                                      /* Automatically merge PRs that build successfully and have this label. */
  // mergifyOptions: undefined,                                                /* Options for mergify. */
  // minNodeVersion: undefined,                                                /* Minimum Node.js version to require via package.json `engines` (inclusive). */
  // npmDistTag: "latest",                                                     /* The dist-tag to use when releasing to npm. */
  // npmignore: undefined,                                                     /* Additional entries to .npmignore. */
  // npmignoreEnabled: true,                                                   /* Defines an .npmignore file. Normally this is only needed for libraries that are packaged as tarballs. */
  // npmRegistry: "registry.npmjs.org",                                        /* The registry url to use when releasing packages. */
  // packageManager: NodePackageManager.YARN,                                  /* The Node Package Manager used to execute scripts. */
  // peerDependencyOptions: undefined,                                         /* Options for `peerDeps`. */
  // peerDeps: [],                                                             /* Peer dependencies for this module. */
  // projenDevDependency: true,                                                /* Indicates of "projen" should be installed as a devDependency. */
  // projenUpgradeAutoMerge: undefined,                                        /* Automatically merge projen upgrade PRs when build passes. */
  // projenUpgradeSchedule: [ "0 6 * * *" ],                                   /* Customize the projenUpgrade schedule in cron expression. */
  // projenUpgradeSecret: undefined,                                           /* Periodically submits a pull request for projen upgrades (executes `yarn projen:upgrade`). */
  // projenVersion: Semver.latest(),                                           /* Version of projen to install. */
  // pullRequestTemplate: true,                                                /* Include a GitHub pull request template. */
  // pullRequestTemplateContents: undefined,                                   /* The contents of the pull request template. */
  // releaseBranches: [ "master" ],                                            /* Branches which trigger a release. */
  // releaseEveryCommit: true,                                                 /* Automatically release new versions every commit to one of branches in `releaseBranches`. */
  // releaseSchedule: undefined,                                               /* CRON schedule to trigger new releases. */
  // releaseToNpm: false,                                                      /* Automatically release to npm when new versions are introduced. */
  // releaseWorkflow: true,                                                    /* Define a GitHub workflow for releasing from "master" when new versions are bumped. */
  // scripts: {},                                                              /* npm scripts to include. */
  // srcdir: "src",                                                            /* Typescript sources directory. */
  // start: true,                                                              /* Defines a `yarn start` interactive experience. */
  // startOptions: undefined,                                                  /* Options for `yarn start`. */
  // testdir: "test",                                                          /* Tests directory. */
  // workflowBootstrapSteps: "yarn install --frozen-lockfile && yarn projen",  /* Workflow steps to use in order to bootstrap this repo. */
  // workflowContainerImage: undefined,                                        /* Container image to use for GitHub workflows. */
  // workflowNodeVersion: undefined,                                           /* The node version to use in GitHub workflows. */
});

project.synth();

```

eladb commented on src/cli/cmds/new.ts:
```diff
@@ -16,6 +16,7 @@ class Command implements yargs.CommandModule {
   public builder(args: yargs.Argv) {
     args.positional('PROJECT-TYPE', { describe: 'optional only when --from is used and there is a single project type in the external module', type: 'string' });
     args.option('synth', { type: 'boolean', default: true, desc: 'Synthesize after creating .projenrc.js' });
+    args.option('removeComments', { type: 'boolean', default: false, desc: 'Synthesize .projenrc.js without commented options' });
```
Avoid negative switches (reverse the logic and the default):


```suggestion
    args.option('comments', { type: 'boolean', default: true, desc: 'Include commented out options in .projenrc.js (use --no-comments to disable)' });
```

This also leaves us an option to allow different styles. For example, I am thinking that in a subsequent PR we can annotate options as "featured" to reduce huge number of commented out options, so we can do something like:

```
--comments=all
--comments=featured
--no-comments
```

But let's leave this for round 2.

Chriscbr replied:
Good ideas, updated

eladb commented on src/inventory.ts:
```diff
@@ -137,29 +143,34 @@ function discoverOptions(jsii: JsiiTypes, fqn: string): ProjectOption[] {
       }
 
+      let typeName;
+      if (prop.type?.primitive) {
```
Add a comment explaining how the type will look based on these heuristics

eladb commented on src/inventory.ts:
```diff
@@ -137,29 +143,34 @@ function discoverOptions(jsii: JsiiTypes, fqn: string): ProjectOption[] {
+        type: typeName,
         switch: propPath.map(p => decamelize(p).replace(/_/g, '-')).join('-'),
         default: defaultValue,
+        isDefaultDescription: isDescription,
```
Why is this needed here? Can't we use `default` later to determine that?

Chriscbr replied:
It could be used, although it requires making a small change to what value gets passed in "default". I've pushed a commit to illustrate this

Chriscbr replied:
There is only one other place in the library using this `default` field, but it is also removing spaces-followed-by-a-trailing-dash in a similar fashion, so my change isn't breaking as far as I can tell:
https://github.com/projen/projen/blob/a654767a7379309744dd73b48d7fd2f5b5bff67e/src/cli/cmds/new.ts#L41

eladb replied:
I am wondering if the correct rendition would be:

```
// copyrightPeriod: undefined /*  DESCRIPTION (default: current year) */
```

Chriscbr replied:
Maybe? I gave it a try - I think the descriptive defaults are helpful in some cases, but seem redundant in other cases. Here are some representative examples:
```
// copyrightPeriod: undefined,                                /* The copyright years to put in the LICENSE file. (default: current year) */
// maxNodeVersion: undefined,                                 /* Minimum node.js version to require via `engines` (inclusive). (default: no max) */
// dependabotOptions: undefined,                              /* Options for dependabot. (default: default options) */
// pullRequestTemplateContents: undefined,                    /* The contents of the pull request template. (default: default content) */
```
That might just be an issue with the actual documentation though. 🤷‍♂️  Let me know which you prefer

eladb replied:
Looks good. Yes, we can edit some of the docs later.

pahud commented:
Was a little bit surprised to see all the comments when I run `npx projen new` today.

It's a nice feature but I personally prefer to disable by default and opt-in with the CLI argument(--comment).

pahud commented on src/awscdk-construct.ts:
```diff
@@ -7,7 +7,7 @@ export interface AwsCdkConstructLibraryOptions extends ConstructLibraryOptions {
    * Minimum target version this library is tested against.
    *
-   * @default 1.60.0
+   * @default "1.60.0"
```
Why need double quote here?

Now I am getting this

<img width="622" alt="圖片" src="https://user-images.githubusercontent.com/278432/97958234-c41fc780-1de7-11eb-8072-b896c3a00303.png">

eladb replied:
This looks like a bug

eladb replied:
Fixed: https://github.com/projen/projen/commit/52c2c9b20cadacf1b62797b0301e428fd653c969

Chriscbr replied:
Ah looks like that was a regression from one of several times I refactored my code, thanks for fixing it. 😄
